### PR TITLE
Fix AddRoleDialog's create button being enabled before entering data

### LINKED ISSUE #5898: Create Role Dialog
1. Create button is active even if the namespace is not selected (causing a nebulous error notification)

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes https://github.com/lensapp/lens/issues/5898

### CODE REVIEW

**Comment by jim-docker** [comment]
This looks rather big for a patch release...

**Comment by Nokel81** [comment]
@jim-docker I don't think so, the vast majority of the lines changed are within the snapshots.

**Comment by jim-docker** [comment]
> @jim-docker I don't think so, the vast majority of the lines changed are within the snapshots.

I filtered out the snapshots first

File: src/renderer/components/button/button.tsx [hunk 1]
```diff
@@ -49,7 +50,11 @@ export const Button = withTooltip((props: ButtonProps) => {
 
   // render as button
   return (
-    <button type="button" {...btnProps}>
+    <button
+      type="button"
+      {...btnProps}
+      data-waiting={`${waiting}`}
```
Comment (aleksfront) [inline src/renderer/components/button/button.tsx]: Is this necessary? Dialog button can be disabled with native `disabled` prop.

File: src/renderer/components/wizard/wizard.tsx [hunk 2]
```diff
@@ -190,7 +190,9 @@ export class WizardStep<D> extends React.Component<WizardStepProps<D>, WizardSte
     if (this.form.noValidate || this.form.checkValidity()) {
       this.next();
     }
-  }, 100);
+  }, 100, {
+    leading: true,
```
Comment (aleksfront) [inline src/renderer/components/wizard/wizard.tsx]: Can you explain what it does?

File: src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx [hunk 3]
```diff
@@ -30,25 +30,32 @@ const clusterFrameLayoutChildComponentInjectable = getInjectable({
...
+                    <Redirect to={startUrl.get()} />
+                  )
+                  : (
+                    <div className={styles.centering}>
+                      <div className="error">
+                        An error has occured. No route can be found matching the
+                        current route, which is also the starting route.
+                      </div>
+                    </div>
+                  )
```
Comment (aleksfront) [inline src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx]: Seems that this changes completely unrelated to what PR states. Should be moved to separate PR.

**Review by aleksfront — CHANGES_REQUESTED** [review]
This PR does more that it states in the title and description. Also, we should find a way to disable a button without new `data-waiting` prop since it changes a huge amount of snapshots. If needed, it can be added to the specified button only.

File: src/renderer/components/wizard/wizard.tsx (hunk 2 above)
Comment (Nokel81) [reply src/renderer/components/wizard/wizard.tsx]: Sure.

This configures how the `debounce`-ed function is called when `this.submit` is called. Namely, it now calls the inner function immediately and then won't call it again for at least another 100ms. Whereas it used to wait for the first 100ms before calling the inner function.

This article (at the given heading and the next heading of "leading edge") explains it very well.

https://css-tricks.com/debouncing-throttling-explained-examples/#aa-debounce 

---

As to why I needed to do this:

It was for getting the unit tests to pass. If the inner function is only called after 100ms then I would have to add that amount of delay into the test (which cases other problems), but with this configuration change the action happens immediately (and subsequent calls within 100ms are then just ignored) makes the tests easier to write and understand.

File: src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx (hunk 3 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx]: Yes probably, I did this so that the `AddRoleDialog` could be its own registered component instead of being tagged on to the `Roles` registered component.

File: src/renderer/components/button/button.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/components/button/button.tsx]: This is not really for having it be disabled but instead being able to write a test for the button to show that it is in the loading state. Similar to other tests that makes sure that a spinner is shown while something is happening.

Comment (Nokel81) [reply src/renderer/components/button/button.tsx]: I will move this change to a separate PR to make it easier to review

File: src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx (hunk 3 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx]: I will move this to a separate PR to make it easier to review

File: src/renderer/components/button/button.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/components/button/button.tsx]: I have made https://github.com/lensapp/lens/pull/6040

File: src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx (hunk 3 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame-layout-child-component.injectable.tsx]: Removed, we can do this in the future.

File: src/renderer/components/wizard/wizard.tsx (hunk 2 above)
Comment (jansav) [reply src/renderer/components/wizard/wizard.tsx]: > It was for getting the unit tests to pass. 

Test shouldn't require this. In case you want to make tests pass without this, then you would have to use `useFakeTime` and `advanceFakeTime`. Here, you are changing behaviour to make test pass, which is wrong.

Comment (Nokel81) [reply src/renderer/components/wizard/wizard.tsx]: It is not wrong if the behaviour was incorrect. The point of this denouncing is to make sure that we don't accidentally double submit.
